okhttp/src/commonMain/kotlin/okhttp3/internal/idn/IdnaMappingTable.kt

```diff
@@ -0,0 +1,265 @@
+ * bits.
+ *
+ * We split each code point into a 14-bit prefix and a 7-bit suffix. All code points with the same
+ * prefix are called a 'section'. There are 128 code points per section.
+ *
+ * Ranges Data (32,612 bytes)
+ * ==========================
+ *
+ * Each entry is 4 bytes, and represents a _range_ of code points that all share a common 14-bit
+ * prefix. Entries are sorted by their complete code points.
+ *
+ * The 4 bytes are named b0, b1, b2 and b3. We also define these supplemental values:
+ *
+ *  * **b2a**: b2 + 0x80
+ *  * **b3a**: b3 + 0x80
+ *  * **b2b3**: (b2 << 7) + b3
+ *
+ * b0
+ * --
+ *
+ * The inclusive start of the range. We get the first 14 bits of this code point from the section
+ * and the last 7 bits from this byte.
+ *
+ * The end of the range is not encoded, but can be inferred by looking at the start of the range
+ * that follows.
+ *
+ * b1
+ * --
+ *
+ * This is either a mapping decision or the length of the mapped output, according to this table:
+ *
+ * ```
+ * 0..63 : Length of the UTF-16 sequence that this range maps to. The offset is b2b3.
+ *   119 : Ignored.
+ *   120 : Valid.
+ *   121 : Disallowed
+ *   122 : Mapped inline to the sequence: [b2].
+ *   123 : Mapped inline to the sequence: [b2a].
+ *   124 : Mapped inline to the sequence: [b2, b3].
+ *   125 : Mapped inline to the sequence: [b2a, b3].
+ *   126 : Mapped inline to the sequence: [b2, b3a].
+ *   127 : Mapped inline to the sequence: [b2a, b3a].
+ *
+ * The range goes until the beginning of the next range.
+ *
+ * When b2 and b3 are unused, their values are set to 0x2d ('-').
+ *
+ * Section Index (1,240 bytes)
+ * ===========================
+ *
+ * Each entry is 4 bytes, and represents all the code points that share a 14-bit prefix. Entries are
+ * sorted by this 14-bit prefix.
+ *
+ * We define these values:
+ *
+ *  * **b0b1s7**: (b0 << 14) + (b1 << 7)
+ *  * **b2b3s2**: (b2 << 9) + (b3 << 2)
+ *
+ * b0b1s7 is the section prefix. If a section is omitted, that means its ranges data exactly matches
+ * that of the preceding section.
+ *
+ * b2b3s2 is the offset into the ranges data. It is shifted by 2 because ranges are 4-byte aligned.
+ *
+ * Mappings Data (4,719 bytes)
+ * ===========================
+ *
+ * This is UTF-8 character data. It is indexed into by b2b3 in the ranges dataset.
+ *
+ * Mappings may overlap.
+ *
+ * ASCII-Only
+ * ==========
+ *
+ * Neither the section index nor the ranges data use bit 0x80 anywhere. That means the data is
+ * strictly ASCII. This is intended to make it efficient to encode this data as a string, and to
+ * index into it as a string.
+ *
+ * The mappings data contains non-ASCII characters.
+ */
+internal class IdnaMappingTable internal constructor(
+  val sections: String,
+  val ranges: String,
+  val mappings: String,
+) {
+  /**
+   * Returns true if the [codePoint] was applied successfully. Returns false if it was disallowed.
+   */
+  fun map(codePoint: Int, sink: BufferedSink): Boolean {
```

Comment:
Should this be on an interface?

Reply:
I guess I'm wondering if we need to load this file in most apps? What does loading a 100KiB class at runtime on android look like?

Would a fast ascii implementation that is used until we hit a non ascii domain be worthwhile? Or I'm overthinking it?

Reply:
A fast ASCII one is a good idea!